File: golang/10_canary_sdk/rollout_test.go
package _0_canary_sdk

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMain(m *testing.M) {
	//val:
	//{
	//    "accesslist": ["hello"],
	//    "blocklist": ["rest"],
	//    "percentage": 50
	//}
	os.Exit(m.Run())
}

func TestCanRollout(t *testing.T) {
	testKey := "test_key"
	type args struct {
		key     string
		methods []Method
	}
	tests := []struct {
		name string
		args args
		want bool
	}{
		{
			name: "invalid key",
			args: args{
				key: "invalid",
				methods: []Method{
					UseAccessListMethod("hello"),
					UsePercentageMethod("test,this"),
				},
			},
			want: false,
		},
		{
			name: "valid key but rollout methods are false",
			args: args{
				key: testKey,
				methods: []Method{
					UseAccessListMethod("invalid"),
					UsePercentageMethod("test,this"),
				},
			},
			want: false,
		},
		{
			name: "valid key and the first rollout method is true",
			args: args{
				key: testKey,
				methods: []Method{
					UseAccessListMethod("hello"),
					UsePercentageMethod("invalid"),
				},
			},
			want: true,
		},
		{
			name: "valid key but first rollout is false but second rollout is true",
			args: args{
				key: testKey,
				methods: []Method{
					UseAccessListMethod("hello"),
					UsePercentageMethod("valid"),
				},
			},
			want: true,
		},
		{
			name: "valid key but middle rollout method must not rollout",
			args: args{
				key: testKey,
				methods: []Method{
					UseAccessListMethod("hello"),
					UseBlockListMethod("rest"),
					UsePercentageMethod("test,this"),
				},
			},
			want: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsAllowed(tt.args.key, tt.args.methods...); got != tt.want {
				t.Errorf("IsAllowed() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestDefaultMethods(t *testing.T) {
	methods := DefaultMethods("stuff")
	assert.Equal(t, 2, len(methods))
}

File: myjava/src/test/java/org/example/canary/RolloutTest.java
package org.example.canary;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class RolloutTest {

    private static final String TEST_KEY = "test_key";

    @BeforeAll
    public static void setup() {
        // Simulate the dynamic configuration retrieval
        // Example configuration:
        // {
        //     "accessList": ["hello"],
        //     "blockList": ["rest"],
        //     "percentage": 50
        // }
        // This setup can involve initializing configurations or other necessary setups.
    }

    @Test
    public void testInvalidKey() {
        boolean result = Rollout.isAllowed("invalid",
                Rollout.useAccessListMethod("hello"),
                Rollout.usePercentageMethod("test,this")
        );
        assertFalse(result, "Expected rollout to be disallowed for invalid key.");
    }

    @Test
    public void testValidKeyButRolloutMethodsAreFalse() {
        boolean result = Rollout.isAllowed(TEST_KEY,
                Rollout.useAccessListMethod("invalid"),
                Rollout.usePercentageMethod("test,this")
        );
        assertFalse(result, "Expected rollout to be disallowed when rollout methods evaluate to false.");
    }

    @Test
    public void testValidKeyAndFirstRolloutMethodIsTrue() {
        boolean result = Rollout.isAllowed(TEST_KEY,
                Rollout.useAccessListMethod("hello"),
                Rollout.usePercentageMethod("invalid")
        );
        assertTrue(result, "Expected rollout to be allowed when the first rollout method evaluates to true.");
    }

    @Test
    public void testValidKeyButFirstRolloutIsFalseSecondIsTrue() {
        boolean result = Rollout.isAllowed(TEST_KEY,
                Rollout.useAccessListMethod("invalid"),
                Rollout.usePercentageMethod("test_key")
        );
        assertTrue(result, "Expected rollout to be allowed when the second rollout method evaluates to true.");
    }

    @Test
    public void testValidKeyButMiddleRolloutMethodMustNotRollout() {
        boolean result = Rollout.isAllowed(TEST_KEY,
                Rollout.useAccessListMethod("hello"),
                Rollout.useBlockListMethod("rest"),
                Rollout.usePercentageMethod("test,this")
        );
        assertFalse(result, "Expected rollout to be disallowed due to block list method.");
    }

    @Test
    public void testDefaultMethods() {
        Rollout.Method[] methods = Rollout.defaultMethods("stuff");
        assertEquals(2, methods.length, "Expected default methods to return two methods.");
    }
}
